### DISCUSSION TITLE
How to configure a NATS connection?

### DISCUSSION THREAD

**Discussion by barakamwakisha** [Netflix/conductor-community#112] [Q&A]
I'm trying to subscribe to messages from a NATS broker.

### What I have done
Added the conductor nats dependency to the `server/build.gradle` file
```gradle
 implementation 'com.netflix.conductor:conductor-nats:3.11.1'
```
Added these properties to the `server/src/main/resources/application.properties` and `docker/server/config/config.properties` files (I'm using Dynomite for persistence)
```
conductor.event-queues.nats.enabled=true
conductor.event-queues.nats.url=nats://0.0.0.0:4222
```

Conductor builds successfully, however, the error below is thrown when I publish a message on NATS
```
ERROR com.netflix.conductor.contribs.queue.nats.NATSAbstractQueue [] - Monitor failed with java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222 for events:complete_add_order
java.lang.RuntimeException: java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222
	at com.netflix.conductor.contribs.queue.nats.NATSObservableQueue.connect(NATSObservableQueue.java:50) ~[conductor-nats-3.11.1.jar!/:3.11.1]
	at com.netflix.conductor.contribs.queue.nats.NATSAbstractQueue.monitor(NATSAbstractQueue.java:249) ~[conductor-nats-3.11.1.jar!/:3.11.1]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222
	at io.nats.client.impl.NatsConnection.connect(NatsConnection.java:207) ~[jnats-2.15.6.jar!/:2.15.6]
	at io.nats.client.impl.NatsImpl.createConnection(NatsImpl.java:29) ~[jnats-2.15.6.jar!/:2.15.6]
	at io.nats.client.Nats.createConnection(Nats.java:276) ~[jnats-2.15.6.jar!/:2.15.6]
	at io.nats.client.Nats.connect(Nats.java:107) ~[jnats-2.15.6.jar!/:2.15.6]
	at com.netflix.conductor.contribs.queue.nats.NATSObservableQueue.connect(NATSObservableQueue.java:45) ~[conductor-nats-3.11.1.jar!/:3.11.1]
```
I'm running Conductor v3.11.1. Is there something I misconfigured?

**Comment by ScottJOster** [comment]
> I'm trying to subscribe to messages from a NATS broker.
> 
> ### What I have done
> Added the conductor nats dependency to the `server/build.gradle` file
> 
> ```groovy-gradle
>  implementation 'com.netflix.conductor:conductor-nats:3.11.1'
> ```
> 
> Added these properties to the `server/src/main/resources/application.properties` and `docker/server/config/config.properties` files (I'm using Dynomite for persistence)
> 
> ```
> conductor.event-queues.nats.enabled=true
> conductor.event-queues.nats.url=nats://0.0.0.0:4222
> ```
> 
> Conductor builds successfully, however, the error below is thrown when I publish a message on NATS
> 
> ```
> ERROR com.netflix.conductor.contribs.queue.nats.NATSAbstractQueue [] - Monitor failed with java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222 for events:complete_add_order
> java.lang.RuntimeException: java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222
> 	at com.netflix.conductor.contribs.queue.nats.NATSObservableQueue.connect(NATSObservableQueue.java:50) ~[conductor-nats-3.11.1.jar!/:3.11.1]
> 	at com.netflix.conductor.contribs.queue.nats.NATSAbstractQueue.monitor(NATSAbstractQueue.java:249) ~[conductor-nats-3.11.1.jar!/:3.11.1]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
> 	at java.util.concurrent.FutureTask.runAndReset(Unknown Source) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
> 	at java.lang.Thread.run(Unknown Source) [?:?]
> Caused by: java.io.IOException: Unable to connect to NATS servers: nats://localhost:4222
> 	at io.nats.client.impl.NatsConnection.connect(NatsConnection.java:207) ~[jnats-2.15.6.jar!/:2.15.6]
> 	at io.nats.client.impl.NatsImpl.createConnection(NatsImpl.java:29) ~[jnats-2.15.6.jar!/:2.15.6]
> 	at io.nats.client.Nats.createConnection(Nats.java:276) ~[jnats-2.15.6.jar!/:2.15.6]
> 	at io.nats.client.Nats.connect(Nats.java:107) ~[jnats-2.15.6.jar!/:2.15.6]
> 	at com.netflix.conductor.contribs.queue.nats.NATSObservableQueue.connect(NATSObservableQueue.java:45) ~[conductor-nats-3.11.1.jar!/:3.11.1]
> ```
> 
> I'm running Conductor v3.11.1. Is there something I misconfigured?

Did you ever get an answer to this I have exactly the same error despite setting new url in my config properties it’s still getting localhost 4222 from somewhere ? how do I set this property (I’m on conductor main build but have added the nats module in gradle ). Thanks

**Reply by barakamwakisha** [reply]
Hi Scott. I ended up using NATS JetStream. Here's how I did it;

 1. Check out the latest release, tagged `v3.13.2` at this point in time. 

 2. Add the nats dependency to the `server/build.gradle` file.
```gradle
implementation 'com.netflix.conductor:conductor-nats:3.11.1'
```

3. Add the configuration below to the `server/src/main/resources/application.properties` file
```
conductor.event-queues.jsm.enabled=true
conductor.event-queues.jsm.pollTimeDuration=1000
conductor.event-queues.jsm.url=nats://0.0.0.0:4222
conductor.event-queues.jsm.durableName=durableName
conductor.event-queues.jsm.listenerQueuePrefix=prefix
conductor.event-queues.jsm.streamStorageType=file
```
